### ISSUE TITLE
compressed_speed_distance handling is broken

### ISSUE THREAD

**Issue by sesse** [dtcooper/python-fitparse#7]
Hi,

compressed_speed_distance is pretty broken, at least on my FR70; the speed is decoded wrong, and the distance is not properly unwrapped. Please take a look at the included patch; it might not be perfect, but it seems to fix the problems okay enough for me.

```
commit 4757d3bbb2c01616c8457d64aadc56403e2a0cf4
Author: Steinar H. Gunderson <sgunderson@bigfoot.com>
Date:   Fri Aug 31 19:46:11 2012 +0200

    Fix handling of compressed_speed_distance.

    First of all, the bit decoding of the speed field was missing a shift,
    leading to messed up values. I corrected it so it's in line with decode.c
    in FIT SDK 4.00. There was also seemingly some attempt at converting from
    km/h to m/s that didn't match anything reasonable (maybe it was an attempt
    at fudging the speed to it would be correct despite the missing shift?),
    so I just deleted it. That function now returns raw values.

    Second, I added decoupling of the two fields into speed and distance;
    the user shouldn't need to care how the field was originally stored.
    compressed_speed_distance is still available as raw values if the user
    needs it for whatever reason (mostly since I didn't bother finding out
    how to delete it), but again, it returns raw values now. (The speed
    and distance fields return the expected decoded m/s and m units.)

    Finally, the code now de-wraps the distance field so it doesn't wrap
    around every 256 meters. Again, decode.c is used as a reference here,
    since I couldn't immediately find the spec for this in the PDFs.

diff --git a/fitparse/base.py b/fitparse/base.py
index 6e11afd..164d902 100644
--- a/fitparse/base.py
+++ b/fitparse/base.py
@@ -58,6 +58,7 @@ class FitFile(object):

         self._last_offset = 0
         self._reference_timestamp = None
+        self._accumulated_distance = 0
         self._global_messages = {}
         self.definitions = []
         self.records = []
@@ -177,6 +178,8 @@ class FitFile(object):

         fields = []
         dynamic_fields = {}
+        raw_speed = None
+        raw_distance = None

         for i, (field, f_size) in enumerate(definition.fields):
             f_raw_data, = self._struct_read(field.type.get_struct_fmt(f_size), definition.arch)
@@ -186,6 +189,9 @@ class FitFile(object):
             if field.name == r.COMPRESSED_TIMESTAMP_FIELD_NAME and \
                field.type.name == r.COMPRESSED_TIMESTAMP_TYPE_NAME:
                 self._reference_timestamp = f_raw_data
+            elif field.name == r.COMPRESSED_SPEED_DISTANCE_FIELD_NAME and \
+                 field.type.name == r.COMPRESSED_SPEED_DISTANCE_TYPE_NAME:
+                raw_speed, raw_distance = bound_field.data

             fields.append(bound_field)

@@ -220,8 +226,16 @@ class FitFile(object):
                 fields.append(r.BoundField(timestamp, ts_field))
                 self._last_offset = header.seconds_offset

-        # XXX -- do compressed speed distance decoding here, similar to compressed ts
-        # ie, inject the fields iff they're in definition.type.fields
+            # XXX -- the conversion back and forth to raw values here is a bit icky.
+            speed_field = definition.type.fields.get(r.SPEED_FIELD_DEF_NUM)
+            if speed_field and raw_speed is not None:
+                fields.append(r.BoundField(raw_speed * 1000. / 100., speed_field))
+
+            if raw_distance is not None:
+                self._accumulated_distance += (raw_distance - self._accumulated_distance) & 0xFFF;
+                distance_field = definition.type.fields.get(r.DISTANCE_FIELD_DEF_NUM)
+                if distance_field:
+                    fields.append(r.BoundField(self._accumulated_distance * 100. / 16., distance_field))

         data = r.DataRecord(header, definition, fields)

diff --git a/fitparse/records.py b/fitparse/records.py
index a0c27d5..5e9beb6 100644
--- a/fitparse/records.py
+++ b/fitparse/records.py
@@ -35,9 +35,13 @@ MESSAGE_DATA = 0
 LITTLE_ENDIAN = 0
 BIG_ENDIAN = 1

+SPEED_FIELD_DEF_NUM = 6
+DISTANCE_FIELD_DEF_NUM = 5
 TIMESTAMP_FIELD_DEF_NUM = 253
 COMPRESSED_TIMESTAMP_FIELD_NAME = 'timestamp'
 COMPRESSED_TIMESTAMP_TYPE_NAME = 'date_time'
+COMPRESSED_SPEED_DISTANCE_FIELD_NAME = 'compressed_speed_distance'
+COMPRESSED_SPEED_DISTANCE_TYPE_NAME = 'record-compressed_speed_distance'

 UNKNOWN_FIELD_NAME = 'unknown'

@@ -299,13 +303,12 @@ _convert_local_date_time = lambda x: datetime.datetime.fromtimestamp(631065600 +
 _convert_bool = lambda x: bool(x)


-# XXX -- untested
 # see FitSDK1_2.zip:c/examples/decode/decode.c lines 121-150 for an example
 def _convert_record_compressed_speed_distance(raw_data):
     first, second, third = (ord(b) for b in raw_data)
-    speed = first + (second & 0b1111)
+    speed = first + ((second & 0b1111) << 8)
     distance = (third << 4) + ((second & 0b11110000) >> 4)
-    return speed / 100. / 1000. * 60. * 60., distance / 16.
+    return speed, distance


 class MessageIndexValue(int):
'''
```

**Comment by pR0Ps** [comment]
@sesse Should be fixed in the master branch now (as of 1cc59e64fe76ec26c6a588037a6e9a0678af714a) so I'm closing this issue. Can you confirm? If it isn't, please re-open and provide a sample FIT file to test against.

**Comment by sesse** [comment]
You're about five years late for me to have any interest in this code anymore, sorry. :-)
